compiler/rustc_session/src/options.rs

```diff
@@ -97,10 +190,37 @@ macro_rules! top_level_options {
                 })*
                 hasher.finish()
             }
+
+            pub fn gather_target_modifiers(&self) -> Vec<TargetModifier> {
+                let mut mods = Vec::<TargetModifier>::new();
+                $({
+                    gather_tmods_top_level!($opt,
+                        &self.$opt, &mut mods, &self.target_modifiers,
+                        [$dep_tracking_marker], [$($tmod),*]);
+                })*
+                mods.sort_by(|a, b| a.opt.cmp(&b.opt));
+                mods
+            }
         }
     );
 }
 
+/// Update if new options group will be added
+#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Copy, Clone, Encodable, Decodable)]
+pub enum OptionsTargetModifiers {
+    CodegenOptions(CodegenOptionsTargetModifiers),
+    UnstableOptions(UnstableOptionsTargetModifiers),
+}
```

Comment:
This should probably be generated like the `Options` struct, just in case any top-level arguments are target modifiers.

Reply:
OptionsTargetModifiers enum is now generated by macro to support substruct option groups. But top-level arguments can't now be target modifiers, because they are processed in a different way, they are not parsed by the same "parse_*" functions system. They are set manually in different places.
Target modifiers are now supported only for codegen/unstable option groups.